### ISSUE TITLE
Emit doc nodes with bracketed text from Java. [phab:D960]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1290]
_Created by **zarko** at 2016-07-12 23:53:58_: 

In line with the C++ indexer, this diff emits doc nodes that contain bracketed, UTF-8 encoded comment text for Java code.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-12 23:57:48_:

We can't use Java 8 features yet.

**Comment by kythe** [comment]
_**schroederc** required changes at 2016-07-12 23:57:48_

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-13 00:05:09_:

  - Remove Java 8 dependency

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-13 18:00:48_:

`new ArrayList<>()` is the new preferred style.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-13 18:00:48_:

`new ArrayList(miniAnchors.size())`

**Comment by kythe** [comment]
_**schroederc** wrote at 2016-07-13 18:00:48_:

Summary?

Also, this change is missing the `MiniAnchor.java` source file.

**Comment by kythe** [comment]
_**schroederc** required changes at 2016-07-13 18:00:48_

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-13 19:24:36_:

  - Missing file
  - Address comments.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-13 21:10:43_:

More Java 8 features to be removed

Test with the following or rebase on top of #1292:
```
bazel test --javacopt '-source 7' //...
```

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-07-13 21:31:36_:

do we want to record this failure in some fashion?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2016-07-13 21:31:36_:

are lambdas not Java 8?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-13 22:11:39_:

This would result in the same `VName` for two copies of the same comment which probably isn't wanted.  Position information could be mixed in to fix this.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-13 23:42:56_:

  - Remove Java 8 dependency
  - Missing file
  - Address comments.
  - More Java 8 dependencies.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-14 17:14:17_:

Why isn't this desirable? Note that the `params` are added to the hash below.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 17:26:06_:

`params` would be the same if the comment is identical.

I just assumed we would always want 1 node per documentation comment.

Likewise, are we going to provide positional information so that UIs can link back to the source location of the documentation?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-14 17:37:40_:

There's not necessarily a 1:1 relationship between `doc` nodes and source-level comments.

`params` would be different if the same comment appeared in a different context, e.g.


```
/**
 * Gets {code name}.
 * return {code name}
 */
String getName() { return name; }
```

That comment is likely to appear in a number of different places, but in each one the `name` field would be different, so the comments would have distinguishable signatures.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

Mark all method overrides with `Override`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

`static`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

`static`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

Type variables should usually be a single capital letter.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

Remove unused import

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

All `com.google` imports are grouped together.

Or... if you want to use the freshest, latest, and "best" style, there are only two import groups: static and non-static.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

Actually... those would be the same since you used `code` rather than `link`.

I suppose it will only matter if we don't want to provide positional information about the documentation comment, but I think we do.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

Comment which anchors would match this condition and why we are filtering them.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 18:00:55_:

```
while (!ends.isEmpty() && ends.peek() == sourcePos) {
```

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-14 19:45:33_:

  - Remove Java 8 dependency
  - Missing file
  - Address comments.
  - More Java 8 dependencies.
  - Address comments.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-14 19:46:22_:

~ magic heuristics ~

Since it's conceivable for multiple comment blocks to contribute to the same `doc` node, this would be better handled with edges, possibly from the documentation anchors we already have.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-07-14 19:46:22_:

This is such a weird rule given how verbose other identifiers are

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 19:54:52_:

But **why** are they negative or empty?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2016-07-14 19:54:52_:

Static group is still first.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2016-07-14 19:54:52_

**Comment by kythe** [comment]
_**zarko** wrote at 2016-07-14 20:06:32_:

  - Address comments.

**Comment by kythe** [comment]
_**zarko** closed via commit db02605cb41c3372818411f3a00b19fb1dff5c37 at 2016-07-14 20:28:14_
